storage/src/index/storage.rs

```diff
@@ -18,63 +24,199 @@ pub struct ValueIterator<'a, V> {
     next: Option<&'a Record<V>>,
 }
 
-// An iterator over all values associated with a translated key, allowing for mutation.
-pub struct MutableValueIterator<'a, V> {
-    next: Option<&'a mut Record<V>>,
-}
-
 impl<V> ValueIterator<'_, V> {
     /// Create a `ValueIterator` that returns no items.
     fn empty() -> Self {
         ValueIterator { next: None }
     }
 }
 
-impl<V> MutableValueIterator<'_, V> {
-    /// Create a `MutableValueIterator` that returns no items.
-    fn empty() -> Self {
-        MutableValueIterator { next: None }
-    }
-}
-
 impl<'a, V> Iterator for ValueIterator<'a, V> {
     type Item = &'a V;
 
     fn next(&mut self) -> Option<Self::Item> {
         match self.next {
             Some(next) => {
                 let loc = &next.value;
-                self.next = next.next.as_deref();
+                match next.next {
+                    Some(ref next_next) => self.next = Some(next_next),
+                    None => self.next = None,
+                }
+                Some(loc)
+            }
+            None => None,
+        }
+    }
+}
+
+enum OccupiedOrVacant<'a, K, V> {
+    Occupied(OccupiedEntry<'a, K, Record<V>>),
+    Vacant(VacantEntry<'a, K, Record<V>>),
+}
+
+/// An iterator over all values associated with a translated key, allowing for mutation of the
+/// current element and insertion of new elements at the front of the list.
+pub struct UpdateValueIterator<'a, K, V> {
+    next: Option<*mut Record<V>>,
+    entry: Option<OccupiedOrVacant<'a, K, V>>,
+    collisions_counter: &'a Counter,
+}
+
+/// UpdateValueIterator must be sendable across threads so it can be held across a journal's read
+/// async boundary.
+unsafe impl<K, V> Send for UpdateValueIterator<'_, K, V> {}
+
+impl<'a, K, V> Iterator for UpdateValueIterator<'a, K, V> {
+    type Item = &'a mut V;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.next {
+            Some(next) => {
+                let current = unsafe { &mut (*next) };
+                let loc = &mut current.value;
+                self.next = current
+                    .next
+                    .as_mut()
+                    .map(|next_next| next_next.as_mut() as *mut _);
                 Some(loc)
             }
             None => None,
         }
     }
 }
 
-impl<'a, V> Iterator for MutableValueIterator<'a, V> {
+impl<K, V> UpdateValueIterator<'_, K, V> {
+    /// Insert a new value at the front of the list.
```

Comment:
I think it would be worth offering some rationale as to why we add at the front of the list (you mentioned LRU-inspiration?).

Reply:
done